### PR TITLE
Split Default Command Handler Logic

### PR DESCRIPTION
**Pull request by OoLunar**
# Summary
This separates the internal logic of the default command handler from a single large method into several smaller methods. The intent is to allow other classes to inherit the default command handler and to easily be able to make minor changes while still be able to keep the default command execution behavior with relative ease.

# Changes
- Rename `CommandExecutor` to `DefaultCommandExecutor`
- Exposed all properties and methods to inheriting classes (`protected`)
- Deeply documented each split method
- Add `ContextCheckFailedData` for failed checks
- Documents the command wrapper IL generator utility files

# Notes
Tested.

### CODE REVIEW

File: DSharpPlus.Commands/Invocation/CommandEmitUtil.cs [hunk 1]
```diff
@@ -19,106 +19,122 @@ internal static class CommandEmitUtil
...
         }
-
         else if (method.ReturnType == typeof(Task))
         {
             return GetTaskFunc(method);
         }
 
+        // This could happen for `void` methods when the user explicitly builds a command tree with them.
+        // We probably don't want to support `void` since the user will never be able to respond to the command...
+        // Unless if it was done through context checks... Hm.
```
Comment (akiraveliara) [inline DSharpPlus.Commands/Invocation/CommandEmitUtil.cs:39]: we shouldn't encourage misuse of context checks in the name of an already immensely questionable code pattern, ideally

File: DSharpPlus.Commands/Invocation/CommandEmitUtil.cs [hunk 2]
```diff
@@ -19,106 +19,122 @@ internal static class CommandEmitUtil
     /// <exception cref="InvalidOperationException">Thrown if the command returns anything but ValueTask and Task.</exception>
     public static Func<object?, object?[], ValueTask> GetCommandInvocationFunc(MethodInfo method, object? target)
     {
-        // we assume this is an anonymous method of any sort
+        // This method is very likely to be an anonymous delegate, which happens to be very slow to invoke.
+        // We're going to take the slow path here and simply do `MethodInfo.Invoke`, for lack of a better way.
```
Comment (akiraveliara) [inline DSharpPlus.Commands/Invocation/CommandEmitUtil.cs:23]: if there's no faster way it's not a slow path :when:

File: DSharpPlus.Commands/Invocation/CommandEmitUtil.cs [hunk 3]
```diff
@@ -19,106 +19,122 @@ internal static class CommandEmitUtil
...
-            il.Emit(OpCodes.Ldc_I4, i);
-            il.Emit(OpCodes.Ldelem_Ref);
+            // Ldarg_1 loads the array of arguments.
+            dynamicMethodIlGenerator.Emit(OpCodes.Ldarg_1);
+
+            // Ldc_I4 loads the index of the current argument.
+            dynamicMethodIlGenerator.Emit(OpCodes.Ldc_I4, i);
+
+            // Ldelem_Ref loads the element at the given index from the array.
+            dynamicMethodIlGenerator.Emit(OpCodes.Ldelem_Ref);
```
Comment (akiraveliara) [inline DSharpPlus.Commands/Invocation/CommandEmitUtil.cs:116]: if you insist on copying the instruction documentation, at least use the proper names - `ldarg.1`, `ldc.i4`, `ldelem.ref` - this goes for the entire file

File: DSharpPlus.Commands/Invocation/CommandEmitUtil.cs [hunk 4]
```diff
@@ -19,106 +19,122 @@ internal static class CommandEmitUtil
...
         {
-            il.Emit(OpCodes.Ldarg_0);
-
-            // not static implies declared within a type
+            // Load the instance (this) onto the stack.
+            dynamicMethodIlGenerator.Emit(OpCodes.Ldarg_0);
             if (method.DeclaringType!.IsValueType)
             {
-                il.Emit(OpCodes.Unbox_Any, method.DeclaringType!);
+                // Change the handling behavior (unbox) of the instance (this) from a reference type to a value type.
```
Comment (akiraveliara) [inline DSharpPlus.Commands/Invocation/CommandEmitUtil.cs:100]: i have no idea what this comment is supposed to mean.

File: DSharpPlus.Commands/Invocation/CommandEmitUtil.cs [hunk 5]
```diff
@@ -19,106 +19,122 @@ internal static class CommandEmitUtil
...
-        ConstructorInfo argumentExceptionCtor = typeof(ArgumentException).GetConstructor([typeof(string), typeof(string)])!;
-
         if (!method.IsStatic)
         {
-            il.Emit(OpCodes.Ldarg_0);
-
-            // not static implies declared within a type
+            // Load the instance (this) onto the stack.
+            dynamicMethodIlGenerator.Emit(OpCodes.Ldarg_0);
             if (method.DeclaringType!.IsValueType)
```
Comment (akiraveliara) [inline DSharpPlus.Commands/Invocation/CommandEmitUtil.cs:98]: removing newlines added for logical separation of logically distinct code is generally ill-advisable

**Review by akiraveliara — CHANGES_REQUESTED** [review]

File: DSharpPlus.Commands/DefaultCommandExecutor.cs [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    {
+        if (!awaitCommandExecution)
+        {
+            _ = WorkerAsync(context);
+        }
+        else
+        {
+            await WorkerAsync(context);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:38]: What guarantee

File: DSharpPlus.Commands/DefaultCommandExecutor.cs [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+
+    /// <summary>
+    /// This method will execute the command provided without any safety checks, context checks or event invocation.
+    /// </summary>
+    /// <param name="context">The context of the command being executed.</param>
+    /// <returns>A tuple containing the command object and any error that occurred during execution. The command object may be null when the delegate is static and is from a static class.</returns>
+    public virtual async ValueTask<(object? CommandObject, Exception? Error)> UnconditionallyExecuteAsync(CommandContext context)
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:45]: Right but will anyone really want to make a shim of this class just to access one method? It feels like more hassle than it's worth. More often than not, it's just gonna be used for a `sudo` command

File: DSharpPlus.Commands/DefaultCommandExecutor.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (akiraveliara) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:38]: not pooling instances

File: DSharpPlus.Commands/DefaultCommandExecutor.cs [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+    {
+        // Keep the command object in scope so it can be accessed after the command has been executed.
+        object? commandObject = null;
+
+        try
+        {
+            // If the class isn't static, we need to create an instance of it.
+            if (!context.Command.Method!.DeclaringType!.IsAbstract || !context.Command.Method.DeclaringType.IsSealed)
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:53]: https://stackoverflow.com/a/1175901/10942966
> `static` classes are declared `abstract` and `sealed` at the IL level. So, you can check `IsAbstract` property to handle both `abstract` classes and `static` classes in one go (for your use case).

However, `abstract` classes are not the only types you can't instantiate directly. You should check for things like interfaces ([without the CoClass attribute](https://stackoverflow.com/questions/1093536/how-does-the-c-compiler-detect-com-types)) and types that don't have a constructor accessible by the calling code.

File: DSharpPlus.Commands/DefaultCommandExecutor.cs [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        {
+            errorMessage = "The number of arguments provided does not match the number of parameters the command expects.";
+            return false;
+        }
+
+        errorMessage = null;
+        return true;
+    }
+
+    protected virtual async ValueTask<IReadOnlyList<ContextCheckFailedData>> ExecuteContextChecksAsync(CommandContext context)
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:115]: What are those

File: DSharpPlus.Commands/DefaultCommandExecutor.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (akiraveliara) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:53]: IsAbstract and IsSealed are very, very different things

File: DSharpPlus.Commands/DefaultCommandExecutor.cs [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+
+        // Add the parent's checks last so we can execute the checks in order.
+        Command? parent = context.Command.Parent;
+        while (parent is not null)
+        {
+            checks.AddRange(parent.Attributes.OfType<ContextCheckAttribute>());
+            parent = parent.Parent;
+        }
+
+        // If there are no checks, we can skip this step.
```
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:128]: Previously unreported bug, thanks for catching it

File: DSharpPlus.Commands/DefaultCommandExecutor.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:53]: Yes they are. However that is how C# chooses to represent static classes

[the remark this answers is not in the document]
Comment (akiraveliara) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:53]: this condition will prevent creating an instance of `sealed class Class`, so...

[the remark this answers is not in the document]
Comment (OoLunar) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:53]: Live testing shows otherwise... Are we sure?

[the remark this answers is not in the document]
Comment (akiraveliara) [reply DSharpPlus.Commands/DefaultCommandExecutor.cs:53]: no, actually

**Review by akiraveliara — APPROVED** [review]

**Review by VelvetToroyashi — APPROVED** [review]
Skimming, as to avoid making comments on framework architecture, so at a glance, looks fine

**Pull request merged**